Sources/SwiftGRPCNIO/GRPCClientChannelHandler.swift

```diff
@@ -0,0 +1,200 @@
+///
+/// Errors are also handled by the channel handler. Promises for the initial metadata and
+/// response (if applicable) are failed with first error received. The status promise is __succeeded__
+/// with the error as a `GRPCStatus`. The stream is also closed and any inbound or outbound messages
+/// are ignored.
+internal class GRPCClientChannelHandler<RequestMessage: Message, ResponseMessage: Message> {
+  internal let initialMetadataPromise: EventLoopPromise<HTTPHeaders>
+  internal let statusPromise: EventLoopPromise<GRPCStatus>
+  internal let responseObserver: ResponseObserver<ResponseMessage>
+
+  /// A promise for a unary response.
+  internal var responsePromise: EventLoopPromise<ResponseMessage>? {
+    guard case .succeedPromise(let promise) = responseObserver else { return nil }
+    return promise
+  }
+
+  /// Promise that the `HTTPRequestHead` has been sent to the network.
+  ///
+  /// If we attempt to close the stream before this has been fulfilled then the program will fatal
+  /// error because of an issue with nghttp2/swift-nio-http2.
+  ///
+  /// Since we need this promise to succeed before we can close the channel, `BaseClientCall` sends
+  /// the request head in `init` which will in turn initialize this promise in `write(ctx:data:promise:)`.
+  ///
+  /// See: https://github.com/apple/swift-nio-http2/issues/39.
+  private var requestHeadSentPromise: EventLoopPromise<Void>!
+
+  private enum InboundState {
+    case expectingHeaders
+    case expectingMessageOrStatus
+    case ignore
+  }
+
+  private enum OutboundState {
+    case expectingHead
+    case expectingMessageOrEnd
+    case ignore
+  }
+
+  private var inboundState: InboundState = .expectingHeaders
+  private var outboundState: OutboundState = .expectingHead
+
+  /// Creates a new `GRPCClientChannelHandler`.
+  ///
+  /// - Parameters:
+  ///   - initialMetadataPromise: a promise to succeed on receiving the initial metadata from the service.
+  ///   - statusPromise: a promise to succeed with the outcome of the call.
+  ///   - responseObserver: an observer for response messages from the server; for unary responses this should
+  ///     be the `succeedPromise` case.
+  internal init(
+    initialMetadataPromise: EventLoopPromise<HTTPHeaders>,
+    statusPromise: EventLoopPromise<GRPCStatus>,
+    responseObserver: ResponseObserver<ResponseMessage>
+  ) {
+    self.initialMetadataPromise = initialMetadataPromise
+    self.statusPromise = statusPromise
+    self.responseObserver = responseObserver
+  }
+
+  /// Observe the given status.
+  ///
+  /// The `status` promise is __succeeded__ with the given status despite `GRPCStatus` being an
+  /// `Error`. If `status.code != .ok` then the initial metadata and response promises (if applicable)
+  /// are failed with the given status.
+  ///
+  /// - Parameter status: the status to observe.
+  internal func observeStatus(_ status: GRPCStatus) {
+    if status.code != .ok {
```

Comment:
Should we move `inboundState` to `.ignore` here?

Reply:
I don't think it matters a great deal either way: `observeStatus` is internal and is always followed by closing the stream which changes the state to `.ignore`. Happy to move it up if you like though.

Reply:
I guess it's fine as is.